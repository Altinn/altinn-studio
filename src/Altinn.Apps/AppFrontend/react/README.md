## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
