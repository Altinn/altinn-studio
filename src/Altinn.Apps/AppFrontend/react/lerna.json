{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
