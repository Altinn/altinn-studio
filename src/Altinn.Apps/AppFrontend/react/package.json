{
	"name": "app-frontend",
	"private": true,
	"scripts": {
		"install-deps": "lerna bootstrap --hoist --ci",
		"no-ci-hoist": "lerna bootstrap --hoist --no-ci",
		"install-deps-with-lock": "lerna bootstrap",
		"test": "lerna exec --concurrency 1 -- npm test",
		"install-all": "lerna exec -- npm ci",
		"clean-lock": "lerna exec -- rm ./package-lock.json",
		"clean-node-modules": "rm -rf ./node_modules && lerna exec --concurrency 1 -- rm -rf ./node_modules",
		"update-packages": "lerna exec --concurrency 1 -- npm update --dev",
		"list-outdated": "lerna run list-outdated"
	},
	"devDependencies": {
		"@typescript-eslint/eslint-plugin": "~4.8.2",
		"@typescript-eslint/parser": "~4.8.2",
		"eslint": "~7.14.0",
		"eslint-config-airbnb": "~18.2.1",
		"eslint-plugin-import": "~2.22.1",
		"eslint-plugin-jsx-a11y": "~6.4.1",
		"eslint-plugin-react": "~7.21.5",
		"eslint-plugin-react-hooks": "^4.2.0",
		"ini": "1.3.5",
		"lerna": "3.22.1"
	}
}
