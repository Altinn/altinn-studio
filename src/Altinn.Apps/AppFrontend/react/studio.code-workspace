{
	"folders": [
		{
			"path": "."
		}
  ],
  "settings": {
    "typescript.surveys.enabled": false,
    "typescript.tsdk": "node_modules/typescript/lib",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
      "javascript",
      "javascriptreact",
      { "language": "typescript", "autoFix": true },
      { "language": "typescriptreact", "autoFix": true}
    ]
  }
}
