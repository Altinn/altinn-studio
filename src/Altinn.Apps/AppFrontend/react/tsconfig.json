{
  "compilerOptions": {
    "outDir": "compiled",
    "module": "commonjs",
    "target": "es5",
    "lib": [
      "es6",
      "dom",
      "es7",
      "esnext"
    ],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "./",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": false,
    "noUnusedLocals": true
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
  ],
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "setupTests.ts",
    "templates",
    "**/setupTests.ts"
  ],
}
