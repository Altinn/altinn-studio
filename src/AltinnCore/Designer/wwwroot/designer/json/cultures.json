{
  "cultures": [
    {
      "id": "af",
      "name": "Afrikaan"
    },
    {
      "id": "sq",
      "name": "Albansk"
    },
    {
      "id": "ar",
      "name": "Arabisk"
    },
    {
      "id": "eu",
      "name": "Baskisk"
    },
    {
      "id": "bg",
      "name": "Bulgarsk"
    },
    {
      "id": "da",
      "name": "Dansk"
    },
    {
      "id": "en",
      "name": "Engelsk"
    },
    {
      "id": "et",
      "name": "Estlandsk"
    },
    {
      "id": "fi",
      "name": "Finsk"
    },
    {
      "id": "fr",
      "name": "Fransk"
    },
    {
      "id": "fo",
      "name": "Færøysk"
    },
    {
      "id": "el",
      "name": "Gresk"
    },
    {
      "id": "gd",
      "name": "Gælisk"
    },
    {
      "id": "he",
      "name": "Hebraisk"
    },
    {
      "id": "hi",
      "name": "Hindi"
    },
    {
      "id": "be",
      "name": "Hviterussisk"
    },
    {
      "id": "id",
      "name": "Indonesisk"
    },
    {
      "id": "ga",
      "name": "Irsk"
    },
    {
      "id": "is",
      "name": "Islandsk"
    },
    {
      "id": "it",
      "name": "Italiensk"
    },
    {
      "id": "ja",
      "name": "Japansk"
    },
    {
      "id": "ji",
      "name": "Jiddisch"
    },
    {
      "id": "ca",
      "name": "Katalansk"
    },
    {
      "id": "zh-CN",
      "name": "Kinesisk"
    },
    {
      "id": "ko",
      "name": "Koreansk"
    },
    {
      "id": "hr",
      "name": "Kroatisk"
    },
    {
      "id": "lv",
      "name": "Latvisk"
    },
    {
      "id": "lt",
      "name": "Litauisk"
    },
    {
      "id": "mk",
      "name": "Makedonsk"
    },
    {
      "id": "ms",
      "name": "Malaysisk"
    },
    {
      "id": "mt",
      "name": "Maltesisk"
    },
    {
      "id": "nl",
      "name": "Nederlandsk"
    },
    {
      "id": "nb",
      "name": "Norsk (bokmål)"
    },
    {
      "id": "nn",
      "name": "Norsk (nynorsk)"
    },
    {
      "id": "fa",
      "name": "Persisk"
    },
    {
      "id": "pl",
      "name": "Polsk"
    },
    {
      "id": "pt",
      "name": "Portugisisk"
    },
    {
      "id": "ro",
      "name": "Rumensk"
    },
    {
      "id": "ru",
      "name": "Russisk"
    },
    {
      "id": "sz",
      "name": "Samisk"
    },
    {
      "id": "sr",
      "name": "Serbisk"
    },
    {
      "id": "sk",
      "name": "Slovakisk"
    },
    {
      "id": "sl",
      "name": "Slovensk"
    },
    {
      "id": "sb",
      "name": "Sorbisk"
    },
    {
      "id": "es",
      "name": "Spansk"
    },
    {
      "id": "sx",
      "name": "Sutu"
    },
    {
      "id": "sv",
      "name": "Svensk"
    },
    {
      "id": "th",
      "name": "Thai"
    },
    {
      "id": "cs",
      "name": "Tsjekkisk"
    },
    {
      "id": "ts",
      "name": "Tsonga"
    },
    {
      "id": "tn",
      "name": "Tswana"
    },
    {
      "id": "tr",
      "name": "Tyrkisk"
    },
    {
      "id": "de",
      "name": "Tysk"
    },
    {
      "id": "uk",
      "name": "Ukrainsk"
    },
    {
      "id": "hu",
      "name": "Ungarsk"
    },
    {
      "id": "ur",
      "name": "Urdu"
    },
    {
      "id": "vi",
      "name": "Vietnamesisk"
    },
    {
      "id": "xh",
      "name": "Xhosa"
    },
    {
      "id": "zu",
      "name": "Zulu"
    }
  ]
}
