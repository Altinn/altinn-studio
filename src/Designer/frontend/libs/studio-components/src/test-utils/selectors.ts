import type { RenderResult } from '@testing-library/react';

export function getRootElementFromContainer<Element extends HTMLElement = HTMLElement>(
  container: RenderResult['container'],
): Element {
  return container.children.item(0) as Element;
}

export function getFirstBySelector<Element extends HTMLElement = HTMLElement>(
  container: RenderResult['container'],
  selector: string,
): Element {
  return container.querySelector(selector) as Element;
}
