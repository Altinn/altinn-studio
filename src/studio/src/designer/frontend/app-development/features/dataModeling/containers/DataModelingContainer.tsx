import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import SchemaEditorApp from '@altinn/schema-editor/SchemaEditorApp';
import { Button, createStyles, Grid, makeStyles } from '@material-ui/core';
import { AddCircleOutline, DeleteOutline } from '@material-ui/icons';
import AltinnPopover from 'app-shared/components/AltinnPopover';
import AltinnPopoverSimple from 'app-shared/components/molecules/AltinnPopoverSimple';
import AltinnInputField from 'app-shared/components/AltinnInputField';
import { getLanguageFromKey, getParsedLanguageFromKey } from 'app-shared/utils/language';
import { SchemaSelect } from '../schemaSelect';
import { deleteDataModel, fetchDataModel, createNewDataModel, saveDataModel, setDataModelName } from '../dataModelingSlice';

function getDataModelTypeName(applicationMetadata: any) {
  if (!applicationMetadata || !applicationMetadata.dataTypes) return undefined;
  const dataTypeWithLogic = applicationMetadata.dataTypes.find((dataType: any) => dataType.appLogic);
  return dataTypeWithLogic?.id;
}

const useStyles = makeStyles(
  createStyles({
    root: {
      marginTop: 24,
      marginLeft: 80,
    },
    schema: {
      marginTop: 4,
    },
    button: {
      margin: 4,
    },
  }),
);

export interface IDataModelingContainerProps {
  language: any;
}

export default function DataModelingContainer(props: IDataModelingContainerProps): JSX.Element {
  const classes = useStyles();
  const dispatch = useDispatch();
  const jsonSchema = useSelector((state: IServiceDevelopmentState) => state.dataModeling.schema);
  const dataModelName = useSelector(
    (state: IServiceDevelopmentState) => getDataModelTypeName(state.applicationMetadataState.applicationMetadata),
  );
  const selectedDataModelName = useSelector((state: IServiceDevelopmentState) => state.dataModeling.modelName);
  const fetchModel = (name: string) => {
    dispatch(setDataModelName({ modelName: name }));
    dispatch(fetchDataModel({}));
  };
  const [deleteButtonAnchor, setDeleteButtonAnchor] = React.useState(null);
  const [createButtonAnchor, setCreateButtonAnchor] = React.useState(null);
  const [newModelName, setNewModelName] = React.useState(null);

  React.useEffect(() => {
    if (dataModelName) {
      fetchModel(dataModelName);
    }
  }, [dispatch, dataModelName]);

  const onSchemaSelected = (id: string, schema: any) => {
    fetchModel(schema.id);
  };
  const onSaveSchema = (schema: any) => {
    dispatch(saveDataModel({ schema }));
  };
  const onCreateClick = (event: any) => {
    setCreateButtonAnchor(event.currentTarget);
  };
  const onNewModelNameChanged = (e: any) => {
    setNewModelName(e.target.value);
  };
  const onCreateConfirmClick = () => {
    if (newModelName && newModelName.length > 0) {
      dispatch(createNewDataModel({ modelName: newModelName }));
    }
    setCreateButtonAnchor(null);
    setNewModelName(null);
  };
  const onDeleteClick = (event: any) => {
    setDeleteButtonAnchor(event.currentTarget);
  };
  const onDeleteConfirmClick = () => {
    dispatch(deleteDataModel());
    setDeleteButtonAnchor(null);
  };
  const onCancelDelete = () => {
    setDeleteButtonAnchor(null);
  };
  const onCancelCreate = () => {
    setCreateButtonAnchor(null);
    setNewModelName(null);
  };

  return (
    <div className={classes.root}>
      <Grid container>
        <Grid item>
          <Button
            id='new-button'
            variant='contained'
            className={classes.button}
            startIcon={<AddCircleOutline />}
            onClick={onCreateClick}
          >
            {getLanguageFromKey('general.create_new', props.language)}
          </Button>
        </Grid>
        <Grid item xs={4}>
          <SchemaSelect
            id='schema' value={selectedDataModelName}
            onChange={onSchemaSelected}
          />
        </Grid>
        <Grid item>
          <Button
            id='delete-button'
            variant='contained'
            className={classes.button}
            startIcon={<DeleteOutline />}
            onClick={onDeleteClick}
          >
            {getLanguageFromKey('general.delete', props.language)}
          </Button>
        </Grid>
        <AltinnPopover
          anchorEl={deleteButtonAnchor}
          anchorOrigin={{ horizontal: 'right', vertical: 'bottom' }}
          btnCancelText={getLanguageFromKey('general.cancel', props.language)}
          descriptionText={getParsedLanguageFromKey('administration.delete_model_confirm', props.language, [selectedDataModelName], true)}
          btnConfirmText={getLanguageFromKey('general.continue', props.language)}
          btnPrimaryId='confirm-delete-button'
          btnClick={onDeleteConfirmClick}
          handleClose={onCancelDelete}
          transformOrigin={{ horizontal: 'right', vertical: 'top' }}
        />
        <AltinnPopoverSimple
          anchorEl={createButtonAnchor}
          handleClose={onCancelCreate}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
        >
          <AltinnInputField
            id='newModelInput'
            placeholder='Name'
            btnText='Ok'
            inputFieldStyling={{ width: '250px' }}
            onChangeFunction={onNewModelNameChanged}
            onBtnClickFunction={onCreateConfirmClick}
          />

        </AltinnPopoverSimple>
      </Grid>
      { selectedDataModelName ?
        <SchemaEditorApp
          language={props.language}
          schema={jsonSchema}
          onSaveSchema={onSaveSchema}
          rootItemId={`#/definitions/${selectedDataModelName}`}
        /> : null}
    </div>
  );
}
