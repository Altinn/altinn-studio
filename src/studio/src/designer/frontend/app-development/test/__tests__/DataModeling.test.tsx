/* tslint:disable:jsx-wrap-multiline */
import 'jest';
import * as React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import DataModelingContainer from '../../features/dataModeling/containers/DataModelingContainer';
import * as testData from '../__testdata__/schemaTestData.json';

describe('DataModeling', () => {
  const language = { administration: {} };
  const initialState = {
    applicationMetadataState: {
      applicationMetadata: {
        dataTypes: [
          {
            id: 'ref-data-as-pdf',
            allowedContentTypes: [
              'application/pdf',
            ],
            maxCount: 0,
            minCount: 0,
          },
          {
            id: 'some-existing-model',
            appLogic: {},
          },
        ],
      },
    },
    dataModeling: {
      schema: testData,
      modelName: 'test',
      saving: false,
    },
  };
  let store: any;
  const dispatchMock = () => Promise.resolve({});
  beforeEach(() => {
    store = configureStore()(initialState);
    store.dispatch = jest.fn(dispatchMock);
  });

  it('fetches model on mount', () => {
    let wrapper: any = null;
    act(() => {
      wrapper = mount(
        <Provider store={store}>
          <DataModelingContainer language={language} />
        </Provider>,
        { context: { store } },
      );
    });
    expect(wrapper).not.toBeNull();
    wrapper.mount();

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'dataModeling/setDataModelName',
      payload: {
        modelName: 'some-existing-model',
      },
    });
    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'dataModeling/fetchDataModel',
      payload: {},
    });
  });

  it('dispatches correctly when clicking new', () => {
    let wrapper: any = null;
    act(() => {
      wrapper = mount(
        <Provider store={store}>
          <DataModelingContainer language={language} />
        </Provider>,
        { context: { store } },
      );
    });
    expect(wrapper).not.toBeNull();

    wrapper.find('#new-button').at(0).simulate('click');
    expect(wrapper.find('input').length).toBe(2);

    wrapper.find('input').last().simulate('change', { target: { value: 'test' } });
    wrapper.find('#newModelInput').find('button').simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'dataModeling/createNewDataModel',
      payload: {
        modelName: 'test',
      },
    });
  });

  it('dispatches correctly when clicking delete', () => {
    let wrapper: any = null;
    act(() => {
      wrapper = mount(
        <Provider store={store}>
          <DataModelingContainer language={language} />
        </Provider>,
        { context: { store } },
      );
    });
    expect(wrapper).not.toBeNull();

    wrapper.find('#delete-button').at(0).simulate('click');
    wrapper.find('#confirm-delete-button').at(0).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'dataModeling/deleteDataModel',
      payload: undefined,
    });
  });

  it('does not dispatch create when name is missing', () => {
    let wrapper: any = null;
    act(() => {
      wrapper = mount(
        <Provider store={store}>
          <DataModelingContainer language={language} />
        </Provider>,
        { context: { store } },
      );
    });
    expect(wrapper).not.toBeNull();
    wrapper.find('#new-button').at(0).simulate('click');

    wrapper.find('#newModelInput').find('button').simulate('click');
    expect(store.dispatch).not.toHaveBeenCalledWith({
      type: 'dataModeling/createNewDataModel',
    });
  });
});
