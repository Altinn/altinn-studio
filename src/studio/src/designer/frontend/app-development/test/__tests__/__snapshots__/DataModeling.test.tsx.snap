// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataModeling Should match snapshot 1`] = `
<div
  class="makeStyles-root-1"
>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3"
        id="new-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
          New
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    >
      <div
        class="makeStyles-select-5 css-2b097c-container"
        id="schema-select-schema"
      >
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class=" css-1uccc91-singleValue"
            >
              test
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-2-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3"
        id="delete-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          Delete
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    id="schema-editor-container"
  >
    <div
      class="makeStyles-root-19"
    >
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
        >
          <div
            class="makeStyles-root-19"
            id="schema-editor"
          >
            <button
              class="makeStyles-button-21"
              type="button"
            >
              Save data model
            </button>
            <ul
              aria-multiselectable="false"
              class="MuiTreeView-root makeStyles-tree-20"
              role="tree"
            >
              <li
                aria-expanded="true"
                class="MuiTreeItem-root Mui-expanded"
                id="properties"
                role="treeitem"
                tabindex="0"
              >
                <div
                  class="MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                  >
                    <div
                      style="padding: 5px 0px 5px 0px;"
                    >
                      <span
                        class="makeStyles-iconContainer-22"
                      >
                        <i
                          class="fa fa-datamodel-properties"
                          style="color: white; text-align: center;"
                        />
                      </span>
                       properties
                    </div>
                  </div>
                </div>
                <ul
                  class="MuiCollapse-container MuiTreeItem-group MuiCollapse-entered"
                  role="group"
                  style="min-height: 0px;"
                >
                  <div
                    class="MuiCollapse-wrapper"
                  >
                    <div
                      class="MuiCollapse-wrapperInner"
                    >
                      <li
                        aria-expanded="false"
                        class="MuiTreeItem-root makeStyles-treeItem-47"
                        id="root-schema-item"
                        role="treeitem"
                        tabindex="-1"
                      >
                        <div
                          class="MuiTreeItem-content"
                        >
                          <div
                            class="MuiTreeItem-iconContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M10 17l5-5-5-5v10z"
                              />
                            </svg>
                          </div>
                          <div
                            class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                          >
                            <div
                              class="makeStyles-labelRoot-39"
                            >
                              <p
                                class="MuiTypography-root makeStyles-label-40 MuiTypography-body1"
                              >
                                <span
                                  class="makeStyles-iconContainer-46"
                                >
                                  <i
                                    class="fa fa-datamodel-object"
                                    style="color: white; text-align: center;"
                                  />
                                </span>
                                 
                                #/properties/melding
                              </p>
                              <button
                                aria-controls="simple-menu"
                                aria-haspopup="true"
                                class="MuiButtonBase-root MuiIconButton-root makeStyles-contextButton-44"
                                id="open-context-menu-button"
                                tabindex="0"
                                type="button"
                              >
                                <span
                                  class="MuiIconButton-label"
                                >
                                  <i
                                    class="fa fa-ellipsismenu"
                                  />
                                </span>
                                <span
                                  class="MuiTouchRipple-root"
                                />
                              </button>
                            </div>
                          </div>
                        </div>
                      </li>
                    </div>
                  </div>
                </ul>
              </li>
              <li
                class="MuiTreeItem-root"
                role="treeitem"
                tabindex="-1"
              >
                <div
                  class="MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  />
                  <div
                    class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                  >
                    info
                  </div>
                </div>
              </li>
              <li
                aria-expanded="false"
                class="MuiTreeItem-root"
                role="treeitem"
                tabindex="-1"
              >
                <div
                  class="MuiTreeItem-content"
                >
                  <div
                    class="MuiTreeItem-iconContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 17l5-5-5-5v10z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                  >
                    definitions
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
        >
          <div
            class="makeStyles-root-35"
          >
            <div>
              <p
                class="no-item-selected"
              >
                No item selected
              </p>
              <hr />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
